include/circt/Dialect/SV/SVOps.h

```diff
@@ -44,44 +45,94 @@ enum class CasePatternBit { Zero = 0, One = 1, AnyX = 2, AnyZ = 3 };
 char getLetter(CasePatternBit bit);
 
 // This is provides convenient access to encode and decode a pattern.
-struct CasePattern {
-  IntegerAttr attr;
+class CasePattern {
 
-  struct DefaultPatternTag {};
+public:
+  enum CasePatternKind { CPK_bit, CPK_enum, CPK_default };
+  CasePattern(CasePatternKind kind) : kind(kind) {}
+  virtual ~CasePattern() {}
+  CasePatternKind getKind() const { return kind; }
+
+  virtual Attribute attr() const = 0;
+
+private:
+  const CasePatternKind kind;
+};
+
+class CaseDefaultPattern : public CasePattern {
+public:
+  using AttrType = mlir::UnitAttr;
+  CaseDefaultPattern(MLIRContext *ctx)
+      : CasePattern(CasePatternKind::CPK_default) {
+    unitAttr = AttrType::get(ctx);
+  }
+
+  Attribute attr() const override { return unitAttr; }
+
+  static bool classof(const CasePattern *S) {
+    return S->getKind() == CPK_default;
+  }
+
+private:
+  // The default pattern is recognized by a UnitAttr. This is needed since we
+  // need to be able to determine the pattern type of a case based on an
+  // attribute attached to the sv.case op.
+  UnitAttr unitAttr;
+};
 
+class CaseBitPattern : public CasePattern {
+public:
   // Return the number of bits in the pattern.
-  size_t getWidth() const { return attr.getValue().getBitWidth() / 2; }
+  size_t getWidth() const { return intAttr.getValue().getBitWidth() / 2; }
 
   /// Return the specified bit, bit 0 is the least significant bit.
   CasePatternBit getBit(size_t bitNumber) const;
 
-  /// Return true if this pattern always matches.
-  bool isDefault() const;
-
   /// Return true if this pattern has an X.
   bool hasX() const;
```

Comment:
Might it be easier to put hasX/Z in the parent class so that you don't have to special case the other two?